core/src/main/java/org/jdbi/v3/core/argument/LoggableArgument.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.argument;
+
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Objects;
+import org.jdbi.v3.core.statement.StatementContext;
+
+/**
+ * A wrapper for {@link Argument}s that enables retrieving (e.g. for logging, see {@link org.jdbi.v3.core.statement.SqlLogger}) of the represented value.
+ */
+public class LoggableArgument implements Argument {
```

Comment:
I'm not sure I understand why this wrapper is necessary.  How is it better than simply going through and ensuring all `Argument` types have an appropriate `toString`?

Reply:
I can look into it again but something about not having access to the value object and Argument being a FunctionalInterface... It wasn't possible to retrofit a proper tostring into Argument without breaking compatibility for implementers.

Reply:
I feel like we'd be better off providing `toString()` implementations in all our argument types.

Reply:
@qualidafial not possible without breaking compatibility afaik. I'm all for it if that's acceptable

Reply:
I think we can find a compromise here: in `SqlStatement.toArgument(Type, Object)`, why don't we just wrap each argument in a simple `Argument` wrapper that delegates to the original argument, but overrides `toString()`?

Reply:
> we'd ugly up the code turning all our Argument lambda expressions into anonymous inner classes

That would solve it on jdbi's end, but users would still end up implementing Argument, oblivious of needing to implement toString if they want logging to work. So we might as well include LoggableArgument to provide an easy solution for both jdbi internals and users and to make the point more clear. The question is just if we want to break compatibility by changing either the `Argument` class or all methods that take an instance of it or not...

> a simple Argument wrapper that delegates to the original argument, but overrides toString()

Like `LoggableArgument` does?

Reply:
We could add a `toString` on `Argument` with a default body that returns some helpful message about needing to override it and a pointer to `LoggableArgument` for convenience? Does that allow us to keep it as a `FunctionalInterface`? That method would be clearly visible to every user implementing Argument, would show up when they test logging, and wouldn't break any compatibility...

Reply:
Did you see my [earlier suggestion](https://github.com/jdbi/jdbi/pull/1081#discussion_r184566588)? This would make it unnecessary for everybody to implement `toString()` in their own argument types.

Reply:
I see a lot of other types of Argument binding though, like bean binding and such. I'm a little too fuzzy on those implementations to be able to tell just like that if anything/what needs to be done about those...

Reply:
Bean binding goes through `NamedParameterFinder` which is a parallel lookup API. All the implementations in core implement `toString()` already.